ffi_internals/src/consumer/consumer_struct.rs

```diff
@@ -198,81 +191,147 @@ extension {}: NativeArrayData {{
+                ffi_args = self.ffi_init_args,
+            )
+        } else {
+            format!(
+                r#"
+    public init(
+        {args}
+    ) {{
+        self.pointer = {ffi_init}(
+            {ffi_args}
+        )
+    }}
+                "#,
+                args = self.consumer_init_args,
+                ffi_init = self.init_fn_name,
+                ffi_args = self.ffi_init_args,
+            )
+        }
+    }
 }
 
-impl ConsumerStruct {
+pub struct CustomConsumerStructInputs<'a> {
+    pub type_name: String,
+    pub required_imports: &'a [Path],
+    pub custom_attributes: &'a CustomAttributes,
+    pub init_fn_name: String,
+    pub init_args: &'a [(Ident, Type)],
+    pub getters: &'a [(Ident, Type)],
+    pub free_fn_name: String,
+    pub clone_fn_name: String,
+}
+
+impl<'a> From<CustomConsumerStructInputs<'a>> for ConsumerStruct {
     /// Returns a `ConsumerStruct` for a type that defines its own custom FFI.
     ///
-    #[must_use]
-    pub fn custom(
-        type_name: String,
-        required_imports: Vec<Path>,
-        init_fn_name: String,
-        init_args: &[(Ident, Type)],
-        getters: &[(Ident, Type)],
-        free_fn_name: String,
-        clone_fn_name: String,
-    ) -> Self {
-        let arg_count = init_args.len();
-        let (consumer_init_args, ffi_init_args) = init_args.iter().enumerate().fold(
+    fn from(inputs: CustomConsumerStructInputs<'_>) -> Self {
+        let arg_count = inputs.init_args.len();
+        let (consumer_init_args, ffi_init_args) = inputs.init_args.iter().enumerate().fold(
             (String::new(), String::new()),
             |mut acc, (index, (i, t))| {
                 // Swift rejects trailing commas on argument lists.
                 let trailing_punctuation = if index < arg_count - 1 { ",\n" } else { "" };
                 // This looks like `foo: Bar,`.
-                let consumer_type =
-                    native_type_data::native_type_data_for_custom(t).consumer_type(None);
+                // TODO: This is where we get the expression: String? for the Unit init. Wrong.
+                let consumer_type = NativeTypeData::from(t).consumer_type(None);
                 acc.0.push_str(&format!(
                     "        {}: {}{}",
                     i.to_string(),
                     consumer_type,
                     trailing_punctuation
                 ));
-                // This looks like `foo.toRust(),`.
+                // It's worth noting here that we always clone when calling an initializer -- the
+                // new Rust instance needs to take ownership of the data because it will be owned by
+                // a new Swift instance whose lifetime is unrelated to the lifetime of the
+                // parameters passed to it.
+                // This looks like `foo.clone(),`.
                 acc.1.push_str(&format!(
-                    "            {}.toRust(){}",
+                    "            {}.clone(){}",
```

Comment:
Not sure if you'd wanna go down this path, but I recently discovered the ability to pass in a width when formating: https://doc.rust-lang.org/std/fmt/index.html#width

Might make it easier to do indentation.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=21ba46faaab434ea6deabe1d1eb504bf

Reply:
Yeah I saw you using that in `agrian-what-rust` and was going to steal the hell out of it 😀  Way more readable format and predictable output.

Reply:
Added https://agrian.atlassian.net/browse/DEV-17287 to handle that during hardening.